### PR TITLE
Corrections of QgsOgrProvider implementaion of GDAL 2.0

### PR DESCRIPTION
**Pull request by mj10777**
This is a summery list of the changes to be made to the QgsOgrProvider class which is based on the Proposal 9 of QGIS Grant Applications - August 2016.

https://qgisblog.files.wordpress.com/2016/09/qgisgrantapplicationssummary-september2016.pdf

Despite the fact that this Proposal did not receive the needed votes to be accepted, the need still exists to correct, what I believe to be, critical flaws in the present implementation.

For this reason I have continued working on this and have prepared a workable solution of the known problems which can be viewed here:

https://github.com/mj10777/QGIS/tree/master_2_ogrprovider

(unfortunately I created this about 2 hours before 'master_2' died, so this should be considered as a 'release-2_18' branch)

---
Main problems:

Problem 01:

The retrieval logic for Ogr-Layers changed between gdal 1.* and gdal 2.*.
This was reported on the 31.03.2015:
https://hub.qgis.org/issues/12479

Unfortunately the suggested changes were not implemented, so that at present, when using gdal 2.* , will not return the same list as gdal 1.*
- this geometries contained in a table that has more than 1 geometry will not turn up (mainly spatialite)

The table containing these geometries will show up, but cannot be loaded.

To make matter worse, corrections (30.06.2016) made to resolve:
https://hub.qgis.org/issues/15168

removed the 'layername' parameter from the sent uri, making it impossible to use 'GetLayerByName', which MUST be used after the first table that contains more than 1 geometry
- GetLayer (with id) can be used up to this point, but not afterwords

To retrieve a Layer, where there are duplicate names in the Datasource (issue: 15168) (mainly in none database Datasources such as KML, GML)
- only GetLayer (with id) can be used to retrieve the layer correctly

So the need exists to determine how the Layer should be opened while reading the whole Datasource in QgsOgrProvider::subLayers()
- if it a duplicate 'layername' : GetLayer (with id)
- otherwise GetLayerByName

--
Problem 02:

Since 2015, new code has been added implementing gdal 2.* specific functions, which will only be included when compiled/built against gdal 2.*.

Thus, when QGIS is complied against gdal 2.*, symbols/functions are added
- that does not exist in gdal 1.*.

Nothing was undertaken to check which gdal version is being used when QGIS is running, causing an 'symbol lookup error' for:
- libqgis_core.so: OGR_F_GetFieldAsInteger64, OGR_F_SetFieldInteger64 and OGR_G_ExportToIsoWkb
- libogrprovider.so: OGR_GT_Flatten, OGR_GT_HasZ and OGR_GT_HasM

causing QGIS to crash.

Since QGIS 2.* will most likely never be depoloyed with a build against gdal 2.* and
QGIS 3.* most likely never be depoloyed with a build against gdal 1.* these problems can be avoided.

Since the use of gdal 1.* could falsify 'M'-Data, the use of gdal 1.* while running should be prevented on a build against gdal 2.* .
When QGIS is built against gdal 2.*, the use of gdal runtime versions where the major version is lower should be prevented.

For this build/runtime gdal versions values have been added to qgis.h and qgis.sip, so that the build/runtime versions can be compaired.
This is done in QGis::ogrRuntimeSupport(), which is called at the application start, setting all values
- GDAL_OGR_RUNTIME_SUPPORTED and GDAL_RASTER_RUNTIME_SUPPORTED
- the function will return true if the gdal runtime major version is >= the build major version [GDAL_OGR_RUNTIME_SUPPORTED]
A correspoinding QGis::gdalRuntimeSupport() function will return  GDAL_RASTER_RUNTIME_SUPPORTED which is set in QGis::ogrRuntimeSupport().
- since qgsgdalproviderbase.cpp contains gdal 2.* specific code, which can cause a 'symbol lookup' error

In the QgsOgrProvider and QgsGdalProviderBase constructors the corresponding functions are called, returning when the result is false.
- the functions that loads the supported Drivers will only do so when these functions return true.

Nothing else seems to be needed to prevent the use of a runtime version that is not supported.

--
Problem 03:

The maintaining of the code in QgsOgrProvider has become difficult.

The need exists to consolidate/isolate crucial functions (such a Open, Sublayer) in a way that
- a backup solution is offered to reliably compleate the task
- build/runtime version specific code (including deprecated functions)

are done in 'Wrapper' functions that deal with these problems once and then correctly.

Sample: the logic to determine the Geometry type is done in multiple places
- getOgrGeomType and in SubLayers (where with gdal 2.* an incorrect result will be returned and 3 '#if defined(GDAL_COMPUTE_VERSION)')

Such tasks should be consolidated in extra functions so that any maintainer of QgsOgrProvider would simply use them without having to deal version specific conditions.

These tasks should be added to the already existing class 'QgsOgrProviderUtils'.

--
Major changes:

QgsOgrProvider::open
- now uses QgsOgrProvider::OGRGetLayerWrapper to open the dataset, replacing the logic of
-- either using GetLayerByName OR GetLayer (with id)

This is done 3 times, but also done in the same way in QgsOgrProvider::repack() and QgsOgrFeatureIterator::QgsOgrFeatureIterator

Based on the sent Url, OGRGetLayerWrapper will attempt to open the Datasource as declared in the Url.
- during subLayers() checking was done if GetLayerByName OR GetLayer (with id) is needed
-- for this the ogr-specific sublayer string will send an extra parameter 'ogrgettype=0|1' with this information

If the given method fails, the other method will be attempted
- if that fails the DataSource will be read calling subLayers()

The sublayer string, that successfully opened the Layer will be stored in the class member mSubLayerString.
- when QgsOgrFeatureIterator later runs it will retrieve this value from its copy of QgsOgrProvider (mSource)
--> and load the layer without further checking
 ogrLayer = QgsOgrProviderUtils::OGRGetSubLayerStringWrapper( mConn->ds, mSource-> mSubLayerString )

The ''ogrgettype=0|1" will be a part of the Uri that will be stored in a QGIS project
- so that when loading a project, no further checking is needed

Since the parameters in the Uri are 'optional', older projects not containing 'ogrgettype=0|1" will not fail
- when changes to the properties of a Project-Layer are made, the 'ogrgettype=0|1" will be appended to the Uri string

QgsSublayersDialog will not show the ogrgettype received from subLayers() in the dialog.
- typedef struct LayerDefinition has been adapted to store the ogrgettype value in ' int getType'

QgisApp::askUserForOGRSublayers
- has been adapted to build the uri with ogrgettype (as well as geometrytype and featurescount)
- layername has also been (re-)added [was removed last June]

--
Status:

All of the above have been implemented.

There are areas where a policy decisions should be made before a pull request should be made.

A python test script has been created (tests/src/python/test_provider_ogr_general.py) to test the different conditions
- together with test data - some of which are the original gdal/ogr autotest databases

Altogether 6 tests
1) Spatialite SpatialView - write support
- a special gdal 2.2.0 version that supports writable SpatialViews runs correctly
-- checks CREATE, UPDATE and DELETE TRIGGERs exist and run correctly 
- other gdal versions report that  writable SpatialViews are not supported
2) Spatialite SpatialTable with more than 1 geometry
- created with gdal autotest/auto/ogr_sql_sqlite.py spatialite_8()
--> tests that the results returned  from gdal 1.* and 2.* are the same
3) Spatialite SpatialTables with XYZM values
- created with gdal autotest/auto/ogr_sql_sqlite.py spatialite_5()
--> checks that expected values fro different Z/M/ZM values are correctly retrieved
---> reports that gdal 1.* does not support M/ZM tables
4) GML 2 MultiPolygon with InternalRings
- created from a historically correct complicated geometry from spatialite Database
-- loops through Polygons, Exterior/InteriorRing calculating area
--> comparing the values with the original spatialite result
--> that the area sum of the Polygons is the same as that as the MultiPolygon
--> that the area of the Polygon + the area of the InteriorRing are the same as the ExteriorRing
5) Sample kmz file showing layers with unicode Chinese characters, some of which a duplicate and some with a geometry field
- included in bug report 15168
-- testing Layers without a geometry definition will not be listed
-- testing that the retrieved layers with duplicate names
--> comparing the amount of points and that the first and last points are different.
6) sqlite Integer64
- created [as 'sqlite_test.db'] and filled with gdal autotest/ogr/ogr_sql_sqlite.py ogr_sqlite_1, 2, 11, 12, 13, 14, 15 and 16
-- checking/updating values in Table 'tpoly' which contains 2 fields with BIGINT (Integer64) 

Windows:
No test have been done with Windows to insure that no 'symbol lookup error' occur.

Tested combinations (linux only):
- original 'release-2_18' (2016-10-25) built with gdal 2.2.0
-- running with gdal 2.2.0: brings expected problems with test for tables that has more than 1 geometry
-- running with gdal 1.11.2: receives (as expected) 'symbol lookup error'

- branch 'master_2_ogrprovider'  built with gdal 2.2.0
-- running with gdal 2.2.0: all test run correctly
-- running with gdal 1.11.2: all test run correctly, reporting that M and Integer64 values are not supported - without 'symbol lookup error' 

- branch 'master_2_ogrprovider'  built with gdal 1.11.2
-- running with gdal 2.2.0: all test run correctly, but 
--> Integer64 values are returned as a string
--> Z/N/ZM Geometries Types are shown and correct values retrieved
-- running with gdal 1.11.2: all test run correctly, reporting that M and Integer64 values are not supported 

Documentation:
The new and changed functions in 'QgsOgrProviderUtils' and 'QgsOgrProvider' have been documented
- explaining 'the reason why' thing are being done in a certain way

--
Policy decisions:

With the exception of  'OGR_G_ExportToIsoWkb' (that contains M value support)
- all other functions that cause a 'symbol lookup error' can be emulated

When using gdal 1.*, the result of the 'Integer64' fields are incorrect
- all values > MAX_INT32 will overflow
So when built with gdal 2.* and running with gdal 1.*
- the reading and writing of 'Integer64' fields should not be permitted to avoid the falsification of data

The same is true for QgsGmlStreamingParser::endElement, where 'OGR_G_ExportToIsoWkb' is used.

For QGIS 2.*: the support for gdal 1.* and 2.* should continue as is.

For QGIS 3.0: gdal 1.* support should be removed.

Reasons:
As gdal evolves, it will become more difficult to 'emulate' OGR functions that do not exist in gdal 1.*
- the inclusion of which would cause a 'symbol lookup error' 
A proper support for M-Values will probably become more important in the future than it is now.

The method created to prevent the use of runtime version that is no longer supported is describe in 'Problem 02'.

### CODE REVIEW

File: src/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -2914,6 +2753,956 @@ void QgsOgrProvider::forceReload()
...
+    default:
+      // Do not use ':', as it will mess with the separator used by QgsSublayersDialog::populateLayers()
+      geom = QString( "Unknown WKB (%1)" ).arg( type );
+  }
+  return geom;
+}
+
+QgsWKBTypes::Type QgsOgrProviderUtils::wkbGeometryTypeFromName( const QString& typeName )
+{ // this should be keep uptodate with gdal/ogr/ogr_core.h (as returned in QgsOgrProvider::ogrWkbGeometryTypeName)
+  if ( typeName == "Point" ) return QgsWKBTypes::Point;
```
Comment (nyalldawson) [inline src/providers/ogr/qgsogrprovider.cpp]: Why not use the methods in QgsWkbTypes here?

Comment (mj10777) [inline src/providers/ogr/qgsogrprovider.cpp]: This was initially intended to replace the OGR types checking
- which caused problems with the different gdal versions (symbol lookup error)
   - so the first goal was to get it working with QgsWkbTypes

While doing so I was indeed wondering if 
- QgsOgrProviderUtils::wkbGeometryTypeName
    - could first call QgsWKBTypes::parseType
and
- QgsOgrProviderUtils::wkbGeometryTypeFromName
    - could first call QgsWKBTypes::displayString( Type type )

and then filter out any types that are known to QGIS but not known to OGR
- there are some types defined but not implemented and thus must be filtered out

I will try this out tomorrow and see if I can optimize this part.

**Comment by rouault** [comment]
See my general observations in https://lists.osgeo.org/pipermail/qgis-developer/2016-October/045602.html 
1) The code is much more complicated than needed due to the unnecessary GDAL runtime vs compile time checks. This are use cases that normally don't happen for proper binary distributions.
2) It is far from being obvious that this is appropriate for the QGIS 2.X series. At least we should have something proposer against master before studying backport.
3) for master, we can probably only focus on GDAL 2. I'd suggest dropping support for GDAL 1.X if that makes things easier
4) The |ogrgettype= parameter is not natural in the URI. I'm surprised not to see a |geometry= parameter with the geometry column name instead, since that's the issue that is attempted to be fixed.

**Comment by nyalldawson** [comment]
I'll echo @rouault's opinions:

- I also don't think this is suitable for backporting to the 2.x branch, and should instead be targeted solely at 3.0. Perhaps a minimal version could be created which just addresses the issues in 2.14 without all the other changes.
- I'm also in favour of making 3.0 depend on GDAL >= 2. See also https://github.com/qgis/qgis3.0_api/issues/41

File: src/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (mj10777) [inline src/providers/ogr/qgsogrprovider.cpp]: The 2 functions have been adapted and tested
- tests for CircularString, CompoundCurve, CurvePolygon, MultiCurve and MultiSurface with testdata
- OGR Types unknown in QGIS are also dealt with (returning uknown)

**Comment by mj10777** [comment]
The main reason for me to start working on this solution was, that I wanted to use the new QGIS geometry functions. Until then I was using the 2.10 version that was adapted to work correctly with the (then) development version of gdal 2.0.0.

The main problem was reported in March 2015:
https://hub.qgis.org/issues/12479

After installing the official version in August (over a **year** after the release of gdal 2.0.0), I was therefore very surprised to see that this problem had not been dealt with. 

For my working scenario QGIS is completely unusable together with a gdal 2.* version.

Any claim that using a gdal 2.* version, 17 months after its initial release,  
> are use cases that normally don't happen for proper binary distributions

- is completely unrealistic

The same is true for the scenario that you must
- downgrade to gdal 1.* **or** wait a good 8 months for QGIS 3

This is a **faulty** implementation that needs correction.

Try to load the 'test' table of 
- gdal_220.autotest.ogr_spatialite_8_2_geometries.sqlite

and try to save, after making a change to the Point geometry
- you will receive an error

The linestring geometry will not even be listed.

This solution resolves this problem as well as the problem reported in bug report 15168
- the solution of which removed the 'layername' completely from the URI that QgsOgrProvider receives

Why @rouault is surprised so see the (re-)implementation of 'layername=' (not 'geometry=') is beyond me
* how else should 'GDALDatasetGetLayerByName' be used if there is no layername to use?
    * which **must** be used with table that contain more than one geometry
    * and up to the change in June was the default use

> ogrgettype= parameter is not natural in the URI

* since the function SubLayers did not look for duplicated layernames
    * where  'GDALDatasetGetLayer' **must** be used
* and in cases where a table contains more than one geometry
    * where  'GDALDatasetGetLayerByName' **must** be used

QgsOgrProvider, which now does **not** query the datasource again
* must know which function must be used to open the Layer that will work in a way that resolve both problems

---
We all agree that for QGIS 3.* only gdal 2 should be used
* hopefully correctly

This solution will, when gdal 1* is being used
* to be precise: when the runtime MAJOR version is < the build MAJOR version
   * that gdal runtime version will not be used

The gdal 2* code does **not** use deprecated functions

The porting of this solution to QGIS 3 will be much simpler
* where everything has already been isolated
    * such as 'OGR_DS_GetLayerByName'
    * so must only be removed in 1 place (QgsOgrProviderUtils::OGRGetLayerNameWrapper)
        * instead of 5 or 6 places (including QgsOgrFeatureIterator and QgsOgrLayerItem)

So this solution not only resolves problems in QGIS 2.*
* but also prepares most of what needed for use in QGIS 3

**Comment by rouault** [comment]
* Regarding the version runtime checks in additions of the compile time checks, I'm not sure if you don't understand what I mean or if you disagree with it. My point is that for normal binary distributions, the GDAL version used at runtime will be the one used at compile time (perhaps with a Z' > Z at runtime with GDAL X.Y.Z version numbering), so all runtime checks are not necessary
* It seems that your URI syntax has both the layerid and layername parameters. This is confusing. One could discuss if we should only use layerid or layername, but using both at the same time doesn't seem to be right. My suggestion to use an extra geometry parameter seems to be the more natural to deal with layers with multiple geometry columns, and it is consistant with other providers (spatialite, PG) deal with such situations
* There's indeed one case that would be addressed is that duplicated layer names at the OGR dataset level. I think those should be fixed in the OGR drivers (for KML I'm aware of the issue. But you also mention GML in https://github.com/qgis/QGIS/pull/3697/files#diff-63b6a1d9be718d66b05c5f314071c51eR2987 and I'm surprised of that) themselves (by appending serial numbers), rather than let this complexity reach QGIS.

**Comment by mj10777** [comment]
The gdal version used at runtime will be the first found in the path or LD_LIBRARY_PATH
* at least in linux

The use of layerid and layername has been in use since at least 2.10
* with layername always being tried first, then layerid
* 'layername' is the geometry name
   * since it used by OGR_DS_GetLayerByName / GDALDatasetGetLayerByName

I did not invent this, that was the way it was and saw no reason to change it
* it is also defined in QgsSublayersDialog::LayerDefinition this way

The problems here is that on a table with more than one geometry
* there is no unique id that can be used for each geometry
* the id returned for that table cannot be used with OGR_DS_GetLayer/GDALDatasetGetLayer

With gdal 1.* each geometry returned a unique id and thus was no problem
* which was probably the reason for the change in June, not taking the change in gdal 2 into account

The other (database) providers don't have these problems because the names ('table(geometry)') combination must be unique

Since GML can manually be created by hand (as KML), I assume that this was possible
* have, however, never found one that was created by another software with duplicate name
   * and I searched for this
* my assumption is that it theatrically possible with non-database sources

**Comment by m-kuhn** [comment]
Superseded by #4341

**Pull request closed without merging**